File: solutions/ruby/2022/02/lib/rock_paper_scissors.rb
# frozen_string_literal: true

module Solutions202202
  module RockPaperScissors
    module Shape
      ROCK = :rock
      PAPER = :paper
      SCISSORS = :scissors
    end

    module Outcome
      WIN = :won
      LOOSE = :lost
      DRAW = :draw
    end

    class Round
      SHAPE_SCORES = {
        Shape::ROCK => 1,
        Shape::PAPER => 2,
        Shape::SCISSORS => 3
      }.freeze

      OUTCOMES = {
        [Shape::SCISSORS, Shape::ROCK] => Outcome::WIN,
        [Shape::PAPER, Shape::SCISSORS] => Outcome::WIN,
        [Shape::ROCK, Shape::PAPER] => Outcome::WIN,
        [Shape::ROCK, Shape::SCISSORS] => Outcome::LOOSE,
        [Shape::SCISSORS, Shape::PAPER] => Outcome::LOOSE,
        [Shape::PAPER, Shape::ROCK] => Outcome::LOOSE
      }.freeze

      OUTCOME_SCORES = {
        Outcome::WIN => 6,
        Outcome::DRAW => 3,
        Outcome::LOOSE => 0
      }.freeze

      attr_reader :opponent, :player

      def initialize(opponent, player)
        @opponent = opponent
        @player = player
      end

      def outcome_score
        OUTCOME_SCORES[outcome]
      end

      def score
        shape_score + outcome_score
      end
    end

    class Game
      def self.from_input(input)
        new(input.split("\n").map { |input_line| round_class.from_input_line(input_line) })
      end

      attr_reader :rounds

      def initialize(rounds)
        @rounds = rounds
      end

      def score
        rounds.sum(&:score)
      end
    end
  end
end

File: solutions/rust/2022/02/src/rock_paper_scissors.rs
pub mod part_1;
pub mod part_2;

use std::str::FromStr;

#[derive(Debug)]
enum Shape {
    Rock,
    Paper,
    Scissors,
}

#[derive(Debug)]
pub enum Outcome {
    Win,
    Loose,
    Draw,
}

#[derive(Debug)]
struct Opponent {
    shape: Shape,
}

impl FromStr for Opponent {
    type Err = String;

    fn from_str(shape_input: &str) -> Result<Self, Self::Err> {
        let shape = match shape_input {
            "A" => Shape::Rock,
            "B" => Shape::Paper,
            "C" => Shape::Scissors,
            _ => return Err(format!("{:?} is not a valid shape", shape_input).to_string()),
        };
        Ok(Self { shape })
    }
}

pub trait Round {
    fn shape_score(&self) -> u32;

    fn outcome(&self) -> Outcome;

    fn outcome_score(&self) -> u32 {
        match &self.outcome() {
            Outcome::Win => 6,
            Outcome::Draw => 3,
            Outcome::Loose => 0,
        }
    }

    fn score(&self) -> u32 {
        &self.shape_score() + &self.outcome_score()
    }
}

pub trait Game {
    type RoundType: Round;

    fn rounds(&self) -> &[Self::RoundType];

    fn score(&self) -> u32 {
        self.rounds().iter().map(|round| round.score()).sum()
    }
}
